### PR TITLE
Fix --port setting

### LINKED ISSUE #13277: Gitea --port rewrites app.ini on it's own, breaks SSH access, API not working & no error logged.
- Gitea version (or commit ref): g26d6c1530
- Git version: 2.17.1
- Operating system: Ubuntu 18.04
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite

## Description
I have updated to [latest nightly build](https://dl.gitea.io/gitea/master) today and since that I cannot pull over SSH anymore suddenly. Has been working for years however.

I removed all old logs and checked log dir, appareantly NO relevant error is logged at all. Not even a authentication attempt, seemingly that point is not reached in code. Also not shows any error in shell. I tried like 5 times to pull in the meantime log-tail is running, the only thing that comes up is about notifications cron.

The git pull says:
```
parse "://:9003/api/internal/serv/command/3/ORGNAME/REPONAME?mode=1&verb=git-upload-pack": missing protocol scheme
```
So gitea somehow tries to connect to http interface, even though I cleary pull over SSH, wut? The port for http listener is 9003.  

Screenshot attached.
Build with issue is:  
```
Powered by Gitea Version: 1.14.0+dev-63-g26d6c1530
gitea-master-linux-amd64  10/22/2020 08:47:08 PM +00:00
```

My old version, nightly build from 19. October has been (no issue / bug not present there):
```
./gitea-2020-10-19 --version
Gitea version 1.14.0+dev-25-g9fe4b7b69
```

## Screenshots
Screenshot_20201023-055106

### PR DESCRIPTION
**Pull request by zeripath**
Unfortunately there was an error in #13195 which set the --port
option before the settings were read. This PR fixes this by
moving applying this option to after the the settings are read

However, on looking further into this code I believe that the setPort
code was slightly odd.

Firstly, it may make sense to run the install page on a different
temporary port to the full system and this should be possible with
a --install-port option.

Secondy, if the --port option is provided we should apply it to both
otherwise there will be unusual behaviour on graceful restart

Thirdly, the documentation for --port says that the setting is
temporary - it should therefore not save its result to the configuration

(This however, does mean that authorized_keys and internal links may
not be correct. - I think we need to discuss this option further.)

Fix #13277

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: docs/content/doc/usage/command-line.en-us.md [hunk 1]
```diff
@@ -40,6 +40,7 @@ Starts the server:
 
 - Options:
     - `--port number`, `-p number`: Port number. Optional. (default: 3000). Overrides configuration file.
```
Comment (gsantner) [inline docs/content/doc/usage/command-line.en-us.md:42]: file >>temporary<<. ?

File: cmd/web.go [hunk 2]
```diff
@@ -190,10 +205,6 @@ func setPort(port string) error {
 		defaultLocalURL += ":" + setting.HTTPPort + "/"
 
 		cfg.Section("server").Key("LOCAL_ROOT_URL").SetValue(defaultLocalURL)
-
-		if err := cfg.SaveTo(setting.CustomConf); err != nil {
-			return fmt.Errorf("Error saving generated JWT Secret to custom config: %v", err)
-		}
```
Comment (zeripath) [inline cmd/web.go]: --port states that it is temporary so it shouldn't write to the ini... 

however, this is going to break gitea serv, admin, manager etc. I just don't know what to do here.

In terms of doing a pure bugfix for the linked Issue only this change should be dropped but it's really hard to know what to do here.

File: docs/content/doc/usage/command-line.en-us.md [hunk 3]
```diff
@@ -40,6 +40,7 @@ Starts the server:
...
+    - `--install-port number`: Port number to run the install page on. Optional. (default: 3000). Overrides configuration file.
```
Comment (lunny) [inline docs/content/doc/usage/command-line.en-us.md:43]: I think this change could be another PR.

Comment (zeripath) [reply docs/content/doc/usage/command-line.en-us.md:43]: It's part of the fix really. 

If you make --port apply both before the install page and after the install page then the PORT option in the install page is irrelevant and it is overwritten.

If you make --port only apply once then if there is a graceful restart after install the port that gitea runs on will change.

If you make --port only apply to the install page then that may change behaviour for other people.

In this case the only solution is provide a separate option for the install page then you can have a consistent solution.

**Review by 6543 — APPROVED** [review]
fix it

File: cmd/web.go [hunk 4]
```diff
@@ -190,10 +205,6 @@ func setPort(port string) error {
 		defaultLocalURL += ":" + setting.HTTPPort + "/"
 
 		cfg.Section("server").Key("LOCAL_ROOT_URL").SetValue(defaultLocalURL)
```
Comment (zeripath) [inline cmd/web.go:207]:
```suggestion
		cfg.Section("server").Key("LOCAL_ROOT_URL").SetValue(defaultLocalURL)
		
		if err := cfg.SaveTo(setting.CustomConf); err != nil {	
			return fmt.Errorf("Error saving generated JWT Secret to custom config: %v", err)	
		}	
```

File: cmd/web.go [hunk 5]
```diff
@@ -190,10 +205,10 @@ func setPort(port string) error {
...
-
-		if err := cfg.SaveTo(setting.CustomConf); err != nil {
-			return fmt.Errorf("Error saving generated JWT Secret to custom config: %v", err)
-		}
+
```
Comment (zeripath) [inline cmd/web.go]:
```suggestion

```

File: cmd/web.go [hunk 6]
```diff
@@ -190,10 +205,10 @@ func setPort(port string) error {
...
 
 		cfg.Section("server").Key("LOCAL_ROOT_URL").SetValue(defaultLocalURL)
-
-		if err := cfg.SaveTo(setting.CustomConf); err != nil {
-			return fmt.Errorf("Error saving generated JWT Secret to custom config: %v", err)
-		}
+		
+		if err := cfg.SaveTo(setting.CustomConf); err != nil {	
+			return fmt.Errorf("Error saving generated JWT Secret to custom config: %v", err)	
+		}
```
Comment (zeripath) [inline cmd/web.go]:
```suggestion
		if err := cfg.SaveTo(setting.CustomConf); err != nil {
			return fmt.Errorf("Error saving generated JWT Secret to custom config: %v", err)
		}
```

**Review by techknowlogick — APPROVED** [review]

**Comment by techknowlogick** [comment]
🚀
